cpp/src/arrow/compute/function_internal.cc

```diff
@@ -108,6 +108,27 @@ Result<std::unique_ptr<FunctionOptions>> DeserializeFunctionOptions(
   return FunctionOptionsFromStructScalar(scalar);
 }
 
+Status CheckAllArrayOrScalar(const std::vector<Datum>& values) {
+  for (const auto& value : values) {
+    if (!value.is_value()) {
+      return Status::TypeError("Tried executing function with non-value type: ",
```

Comment:
"Value type" is a bit cryptic TBH. Can we reword this more explicitly?
```suggestion
      return Status::TypeError("Tried executing function with non-array, non-scalar type: ",
```